### PR TITLE
3. Improve handling of qr timeout & state errors when polling

### PR DESCRIPTION
**Pull request by MKodde**
Improve handling of qr timeout

Prior to this change, the application would not check if the qr code was expired. If the user scanned an expired qr code, it would try to see if it was invalid.
This resulted in errors in the logs, because it should not even try to see if the scanned qr code is valid if it is expired.
This change checks if the qr code is expired before attempting to check its validity.

See https://www.pivotaltracker.com/n/projects/1163646/stories/188205272

---

Improve handling of state errors when polling

Prior to this change, when the code expired, the user was shown the error page. 
This change shows the user a friendly message and a retry link that serves a fresh qr code.
When a state error occurs, an invalid-request message is sent to the status page, that will then show a 'something went wrong, refresh to try again' message.

See https://www.pivotaltracker.com/n/projects/1163646/stories/188205289

### CODE REVIEW

File: src/Controller/AuthenticationQrController.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    #[Route(path: '/authentication/qr', name: 'app_identity_authentication_qr', methods: ['GET'])]
+    #[RequiresActiveSession]
+    public function __invoke(): Response
+    {
+        $nameId = $this->authenticationService->getNameId();
+        $sari = $this->stateHandler->getRequestId();
+        $logger = WithContextLogger::from($this->logger, ['nameId' => $nameId, 'sari' => $sari]);
+        $logger->info('Client request QR image');
+
+        // Do have a valid sample AuthnRequest?.
```
Comment (johanib) [inline src/Controller/AuthenticationQrController.php:56]: 'sample'?

File: src/Controller/AuthenticationQrController.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        if (!$this->authenticationService->authenticationRequired()) {
+            $logger->error('There is no pending authentication request from SP');
+
+            return new Response('No authentication required', Response::HTTP_BAD_REQUEST);
```
Comment (johanib) [inline src/Controller/AuthenticationQrController.php:60]: Johan: To test manually?

**Review by johanib — COMMENTED** [review]
Preliminary check

File: src/EventSubscriber/SessionStateListener.php [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            if ($sessionId !== $sessionCookieId) {
+                $logger->error('The session cookie does not match the session id.');
+                return;
+            }
+        } catch (SessionNotFoundException) {
+            $logger->info('Session not found.');
+            return;
+        }
+
+        $logger->info('User session matches the session cookie.');
```
Comment (johanib) [inline src/EventSubscriber/SessionStateListener.php:81]: Do we want to log this?

File: src/EventSubscriber/SessionStateListener.php [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            'route' => $event->getRequest()->getRequestUri(),
+        ]);
+
+        $sessionCookieId = $event->getRequest()->cookies->get($this->sessionName);
+        if ($sessionCookieId === null) {
+            $logger->info('User made a request without a session cookie.');
+            return;
+        }
+
+        $logger->info('User made a request with a session cookie.');
```
Comment (johanib) [inline src/EventSubscriber/SessionStateListener.php:66]: Do we want to log this?

**Comment by johanib** [comment]
@MKodde I went through this. 

It appears to be in order. Afaiks, all tickets are covered:

https://www.pivotaltracker.com/n/projects/1163646/stories/188205232
https://www.pivotaltracker.com/n/projects/1163646/stories/188205272
https://www.pivotaltracker.com/n/projects/1163646/stories/188205214
https://www.pivotaltracker.com/n/projects/1163646/stories/188205289

I did not test the timeout feature. Is that easy to test on dev?

As for what next, I want to address the issues I found and then squash & write an adequate commit message.
Then we can merge and move to testing.

File: config/packages/framework.yaml [hunk 5]
```diff
@@ -14,8 +14,9 @@ framework:
     # Remove or comment this section to explicitly disable session support.
     session:
         handler_id: null
-        cookie_secure: auto
-        cookie_samesite: lax
+        name: sess_tiqr
```
[the remark this answers is not in the document]
Comment (johanib) [reply config/packages/framework.yaml:17]: discussed: This is probably how it is on prod now. 
Also, tiqr sessions are short lived, so it is not a problem if the sessions get lost (probably).

File: assets/typescript/RegistrationStateMachine.ts [hunk 6]
```diff
@@ -62,6 +64,12 @@ export class RegistrationStateMachine {
         this.qrCode.hide();
         document.location.replace(this.finalizedUrl);
         break;
+      case RegistrationStateMachine.TIMEOUT:
+        this.qrCode.hide();
+        this.statusUi.showTimeoutHappened();
+        this.statusPollingService.stop();
+        this.previousStatus = RegistrationStateMachine.ERROR;
```
Comment (johanib) [inline assets/typescript/RegistrationStateMachine.ts:71]: What is tiqr-server?

File: src/Controller/RegistrationController.php [hunk 7]
```diff
@@ -90,16 +95,24 @@ public function registration(Request $request): Response
...
-        // Do have a valid sample AuthnRequest?.
+
+        // Do we have a valid GSSP registration AuthnRequest in this session?
         if (!$this->registrationService->registrationRequired()) {
             $this->logger->error('There is no pending registration request');
 
             return new Response('No registration required', Response::HTTP_BAD_REQUEST);
         }
+
+        if ($this->tiqrService->isEnrollmentTimedOut()) {
```
Comment (johanib) [inline src/Controller/RegistrationController.php:111]: Prior to this change, I assume the timeouts would occur, but the tiqr 'frontend' would be unaware?

File: assets/typescript/RegistrationStateMachine.ts (hunk 6 above)
Comment (pmeulen) [reply assets/typescript/RegistrationStateMachine.ts:71]: Tiqr-server is the role that Stepup-tiqr has (with help from https://github.com/Tiqr/tiqr-server-libphp). The tiqr-client is the iOS/Android app running on the user's phone.

File: config/packages/framework.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (johanib) [reply config/packages/framework.yaml:17]: cookie_samesite: none 
Should be the value. This is also on prod

**Review by pmeulen — CHANGES_REQUESTED** [review]
- Don't us the app secret when determining the correlation ID

The PR is hard to review as it is now because it implements multiple stories and includes refactoring. Is splitting this up reasonably possible?

File: assets/typescript/RegistrationStateMachine.ts (hunk 6 above)
Comment (pmeulen) [reply assets/typescript/RegistrationStateMachine.ts:71]: So a "tiqr-server" it is (part of) what Stepup-tiqr is, it not a separate server or process from Stepup-tiqr.

**Comment by johanib** [comment]
> * Don't use the app secret when determining the correlation ID
> 
> 
> The PR is hard to review as it is now because it implements multiple stories and includes refactoring. Is splitting this up reasonably possible?

Hi Pieter, 

I split the pr's up in smaller parts, also splitting the refactors into separate pr's, to be reviewed in order:

* https://github.com/OpenConext/Stepup-tiqr/pull/224
* https://github.com/OpenConext/Stepup-tiqr/pull/225
* https://github.com/OpenConext/Stepup-tiqr/pull/210
* https://github.com/OpenConext/Stepup-tiqr/pull/226
* https://github.com/OpenConext/Stepup-tiqr/pull/227

I hope this makes it easier to review.

**Comment by MKodde** [comment]
Given I'm the author of this PR, I'm not able to approve this change. But for what its worth :heavy_check_mark:

File: src/Service/SessionCorrelationIdService.php [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+    public function generateCorrelationId(): ?string
+    {
+        $sessionCookie = $this->requestStack->getMainRequest()?->cookies->get($this->sessionName);
+
+        if ($sessionCookie === null) {
+            return null;
+        }
+
+        return hash('sha256', $this->appSecret . substr($sessionCookie, 0, 10));
```
[the remark this answers is not in the document]
Comment (MKodde) [reply src/Service/SessionCorrelationIdService.php]: This was fixed in #226

**Review by pmeulen — APPROVED** [review]

**Pull request merged**